### PR TITLE
Fixes issue with async / await steps not being reported correctly

### PR DESCRIPTION
**Pull request by cosminstirbu**
Currently async steps are reported twice.

We've used this fix for a while in our project (via a SpecNuts fork) and we'd like to merge it back in the main repository and maybe plan a NuGet release as well.

### CODE REVIEW

File: SpecNuts/Reporters.cs [hunk 1]
```diff
@@ -96,15 +99,15 @@ internal static Step CreateStep(DateTime starttime, MethodBase method, params ob
...
-							    step.Rows = new List<Row> {new Row() {Cells = table.Header.ToList()}};
-
-                                foreach (var tableRow in table.Rows)
-                                {
-                                    step.Rows.Add(new Row()
-                                    {
-                                        Cells = tableRow.Select(x => x.Value).ToList()
-                                    });
-                                }
+								step.Rows = new List<Row> { new Row() { Cells = table.Header.ToList() } };
```
Comment (cosminstirbu) [inline SpecNuts/Reporters.cs]: Formatting changes, not relevant to the actual fix.

File: SpecNuts/Reporters.cs [hunk 2]
```diff
@@ -58,13 +59,15 @@ internal static Step CreateStep(DateTime starttime, MethodBase method, params ob
 						if (table != null)
 						{
 
-						    step.Rows = new List<Row> {new Row() {Cells = table.Header.ToList()}};
+							step.Rows = new List<Row> { new Row() { Cells = table.Header.ToList() } };
```
Comment (cosminstirbu) [inline SpecNuts/Reporters.cs]: Formatting changes, not relevant to the actual fix.

File: SpecNuts/Reporters.cs [hunk 3]
```diff
@@ -39,9 +40,9 @@ internal static Step CreateStep(DateTime starttime, MethodBase method, params ob
 			{
 				Name = ScenarioStepContext.Current.StepInfo.Text,
 				StartTime = starttime,
-                Keyword = ScenarioContext.Current.CurrentScenarioBlock + " ",
-                Id = ScenarioStepContext.Current.StepInfo.Text.Replace(" ", "-").ToLower()
-            };
+				Keyword = ScenarioContext.Current.CurrentScenarioBlock + " ",
```
Comment (cosminstirbu) [inline SpecNuts/Reporters.cs]: Formatting changes, not relevant to the actual fix.

File: SpecNuts/Reporters.cs [hunk 4]
```diff
@@ -129,19 +132,19 @@ internal static Step CreateStep(DateTime starttime, MethodBase method, params ob
...
 			}
 
-		    step.Name = ScenarioStepContext.Current.StepInfo.Text;
+			step.Name = ScenarioStepContext.Current.StepInfo.Text;
 
 			return step;
 		}
 
-		internal static void ExecuteStep(Action action, params object[] args)
+		internal static async Task ExecuteStep(Func<Task> stepFunc, params object[] args)
```
Comment (cosminstirbu) [inline SpecNuts/Reporters.cs]: As part of the fix we've changed this to a Task

File: SpecNuts/Reporters.cs [hunk 5]
```diff
@@ -201,21 +204,26 @@ internal static void ExecuteStep(Action action, MethodBase methodBase, params ob
 				}
 
 
-                foreach (var reporter in GetAll())
+				foreach (var reporter in GetAll())
```
Comment (cosminstirbu) [inline SpecNuts/Reporters.cs]: Formatting changes, not relevant to the actual fix.

File: SpecNuts/ReportingAspect/ReportingMessageSink.cs [hunk 6]
```diff
@@ -19,26 +20,40 @@ public IMessageCtrl AsyncProcessMessage(IMessage msg, IMessageSink replySink)
...
+			// Avoid reporting proxy method calls for fields injected via the constructor
+			// in the step definition classes. This caused some steps to pe reported twice.
+			if (methodMessage.MethodName == "FieldGetter")
+			{
+				return NextSink.SyncProcessMessage(msg);
+			}
 
-			Reporters.ExecuteStep(
+			IMethodReturnMessage mrm = null;
+			var task = Reporters.ExecuteStep(
```
Comment (cosminstirbu) [inline SpecNuts/ReportingAspect/ReportingMessageSink.cs]: The fix consists in wrapping the step definition task with a task that allows reporting the execution of that step.

Please note that if the return type was not a task, we're passing to the ExecuteStep method a completed Task, so it doesn't impact the step definitions that return void.

File: SpecNuts/SpecNuts.csproj [hunk 7]
```diff
@@ -59,20 +59,21 @@
...
     <Reference Include="System.Core" />
     <Reference Include="System.Xml.Linq" />
     <Reference Include="System.Data.DataSetExtensions" />
     <Reference Include="Microsoft.CSharp" />
     <Reference Include="System.Data" />
     <Reference Include="System.Xml" />
-    <Reference Include="TechTalk.SpecFlow, Version=2.1.0.0, Culture=neutral, PublicKeyToken=0778194805d6db41, processorArchitecture=MSIL">
-      <HintPath>..\packages\SpecFlow.2.1.0\lib\net45\TechTalk.SpecFlow.dll</HintPath>
-      <Private>True</Private>
+    <Reference Include="Newtonsoft.Json">
```
Comment (cosminstirbu) [inline SpecNuts/SpecNuts.csproj]: Updated dependencies

**Comment by williamsia** [comment]
@cosminstirbu thanks for all your bug fixes and improvement mate. Will try to review it this week when I have a free time. Also currently, I configured a very simple appveyor project https://ci.appveyor.com/project/williamsia/specresults to push the nuget package, so once it's merged and pass all unit tests, the new nuget package will be published.

**Comment by cosminstirbu** [comment]
@williamsia - Thank you as well for maintaining this. Coming from an "original" Cucumber background, this project makes Specflow reporting much easier.

It's not a blocker for us, so whenever you find time to review it that will be fine.

Regards,
Cosmin

**Pull request merged**